File: js/signal/src/SignalRoom.ts
import type { RoomContext } from '@nolag/js-sdk';
import { EventEmitter } from './EventEmitter';
import { PeerManager } from './PeerManager';
import { generateId, filterEmitOptions, mergeFilters, withoutFilters } from './utils';
import { TOPIC_SIGNALING } from './constants';
import type {
  SignalRoomEvents,
  SignalMessage,
  SignalType,
  Peer,
  SignalPresenceData,
  ResolvedSignalOptions,
  SignalOptions,
  FilterValue,
} from './types';

/**
 * SignalRoom — a single signaling room for WebRTC peer discovery and exchange.
 *
 * Created via `NoLagSignal.joinRoom(name)`. Do not instantiate directly.
 */
export class SignalRoom extends EventEmitter<SignalRoomEvents> {
  /** Room name */
  readonly name: string;

  private _roomContext: RoomContext;
  private _localPeer: Peer;
  private _options: ResolvedSignalOptions;
  private _peerManager: PeerManager;
  private _log: (...args: unknown[]) => void;
  private _isConnected: () => boolean;

  // Stored topic handler ref — cleanup removes exactly this, never all
  // handlers for a topic (the client may be shared with other consumers).
  private _onSignalingRef: ((data: unknown) => void) | null = null;

  /** Filter values applied to the signaling subscription. */
  private _filters: FilterValue[] = [];

  /** @internal */
  constructor(
    name: string,
    roomContext: RoomContext,
    localPeer: Peer,
    options: ResolvedSignalOptions,
    log: (...args: unknown[]) => void,
    isConnected: () => boolean,
  ) {
    super();
    this.name = name;
    this._roomContext = roomContext;
    this._localPeer = localPeer;
    this._options = options;
    this._log = log;
    this._isConnected = isConnected;

    this._peerManager = new PeerManager(localPeer.actorTokenId);
  }

  // ============ Public Properties ============

  /** All remote peers currently in this room */
  get peers(): Map<string, Peer> {
    return this._peerManager.peers;
  }

  // ============ Signaling ============

  /**
   * Send an SDP offer to a specific peer.
   */
  sendOffer(toPeerId: string, offer: RTCSessionDescriptionInit, opts?: SignalOptions): void {
    this.signal(toPeerId, 'offer', offer, opts);
  }

  /**
   * Send an SDP answer to a specific peer.
   */
  sendAnswer(toPeerId: string, answer: RTCSessionDescriptionInit, opts?: SignalOptions): void {
    this.signal(toPeerId, 'answer', answer, opts);
  }

  /**
   * Send an ICE candidate to a specific peer.
   */
  sendIceCandidate(toPeerId: string, candidate: RTCIceCandidateInit, opts?: SignalOptions): void {
    this.signal(toPeerId, 'ice-candidate', candidate, opts);
  }

  /**
   * Send a bye signal to a specific peer (graceful close).
   */
  sendBye(toPeerId: string, opts?: SignalOptions): void {
    this.signal(toPeerId, 'bye', {}, opts);
  }

  /**
   * Send a generic signal message to a specific peer.
   *
   * By default this broadcasts to the room and peers discard messages not
   * addressed to them. Pass `{ filter: toPeerId }` to have the server do the
   * addressing instead, so the signal is only delivered to that peer.
   */
  signal(
    toPeerId: string,
    type: SignalType,
    payload: RTCSessionDescriptionInit | RTCIceCandidateInit | Record<string, unknown>,
    opts?: SignalOptions,
  ): void {
    const message: SignalMessage = {
      id: generateId(),
      type,
      fromPeerId: this._localPeer.peerId,
      toPeerId,
      payload,
      timestamp: Date.now(),
    };

    this._log('Sending signal:', type, '→', toPeerId);

    this._roomContext.emit(TOPIC_SIGNALING, message, { echo: false, ...filterEmitOptions(opts) });
  }

  // ============ Filters ============

  /** This peer's own id — the value to filter on to receive directed signals. */
  get localPeerId(): string {
    return this._localPeer.peerId;
  }

  /** The filter values currently applied to this room's signaling. */
  get filters(): FilterValue[] {
    return [...this._filters];
  }

  /**
   * Replace this room's signaling filters — only signals published with one of
   * these values are delivered. Set your own peerId to receive only signals
   * addressed to you, and send with `{ filter: toPeerId }` so the server does
   * the addressing rather than every peer discarding other peers' traffic.
   *
   * A filtered peer stops receiving unfiltered room broadcasts, so switch the
   * whole room over together. Passing an empty array restores the wildcard
   * subscription, which receives everything.
   *
   * @example
   * ```ts
   * room.setFilters([room.localPeerId]);  // only signals addressed to me
   * room.setFilters([]);                  // back to room broadcast
   * ```
   */
  setFilters(values: FilterValue[]): void {
    this._filters = [...values];
    // The core types filters as `string[]`, but both its implementation and
    // the wire protocol accept AND groups (nested arrays).
    this._roomContext.setFilters(TOPIC_SIGNALING, this._filters as unknown as string[]);
  }

  /** Add filter values to the existing set. Existing AND groups are kept. */
  addFilters(values: string[]): void {
    this.setFilters(mergeFilters(this._filters, values));
  }

  /**
   * Remove filter values from the existing set. Removing the last value
   * restores the wildcard subscription.
   */
  removeFilters(values: string[]): void {
    this.setFilters(withoutFilters(this._filters, values));
  }

  // ============ Peers ============

  /**
   * Get all remote peers in this room.
   */
  getPeers(): Peer[] {
    return this._peerManager.getAll();
  }

  /**
   * Get a specific peer by peerId.
   */
  getPeer(peerId: string): Peer | undefined {
    return this._peerManager.getPeer(peerId);
  }

  // ============ Internal (called by NoLagSignal) ============

  /** @internal Subscribe to signaling topic and attach listeners */
  _subscribe(filters?: FilterValue[]): void {
    this._log('Room subscribe:', this.name);

    this._filters = filters ? [...filters] : [];

    if (this._filters.length > 0) {
      this._roomContext.subscribe(TOPIC_SIGNALING, { filters: this._filters });
    } else {
      this._roomContext.subscribe(TOPIC_SIGNALING);
    }

    // Listen for signals (ref stored for handler-specific removal)
    this._onSignalingRef = (data: unknown) => {
      this._handleIncomingSignal(data);
    };
    this._roomContext.on(TOPIC_SIGNALING, this._onSignalingRef);
  }

  /** @internal Set presence and fetch room members */
  _activate(): void {
    this._log('Room activate:', this.name);
    this._setPresence();

    this._roomContext.fetchPresence().then((actors) => {
      this._log('Room presence fetched:', this.name, actors.length, 'actors');
      for (const actor of actors) {
        if (actor.presence) {
          const peer = this._peerManager.addFromPresence(
            actor.actorTokenId,
            actor.presence as SignalPresenceData,
            actor.joinedAt,
          );
          if (peer) {
            this.emit('peerJoined', peer);
          }
        }
      }
    }).catch((err) => {
      this._log('Failed to fetch room presence:', err);
    });
  }

  /** @internal Re-set presence after reconnect */
  _updateLocalPresence(): void {
    this._setPresence();
  }

  /** @internal Handle a presence:join event */
  _handlePresenceJoin(actorTokenId: string, presenceData: SignalPresenceData): void {
    const peer = this._peerManager.addFromPresence(actorTokenId, presenceData);
    if (peer) {
      this._log('Peer joined room:', this.name, peer.peerId);
      this.emit('peerJoined', peer);
    }
  }

  /** @internal Handle a presence:leave event */
  _handlePresenceLeave(actorTokenId: string): void {
    const peer = this._peerManager.removeByActorId(actorTokenId);
    if (peer) {
      this._log('Peer left room:', this.name, peer.peerId);
      this.emit('peerLeft', peer);
    }
  }

  /** @internal Handle a presence:update event */
  _handlePresenceUpdate(actorTokenId: string, presenceData: SignalPresenceData): void {
    this._peerManager.addFromPresence(actorTokenId, presenceData);
  }

  /** @internal Unsubscribe and clean up */
  _cleanup(): void {
    this._log('Room cleanup:', this.name);

    // Server unsubscribes need a live socket; skip when disconnected
    // (best-effort — the core would no-op with an error callback anyway).
    if (this._isConnected()) {
      this._roomContext.unsubscribe(TOPIC_SIGNALING);
    }

    // Handler-specific removal only: the client may be shared, and a bare
    // off(topic) would strip other consumers' handlers too.
    if (this._onSignalingRef) this._roomContext.off(TOPIC_SIGNALING, this._onSignalingRef);
    this._onSignalingRef = null;

    this._peerManager.clear();
    this.removeAllListeners();
  }

  // ============ Private ============

  private _handleIncomingSignal(data: unknown): void {
    const message = data as SignalMessage;

    // Only process messages targeted at this peer
    if (message.toPeerId !== this._localPeer.peerId) return;

    this._log('Received signal:', message.type, 'from', message.fromPeerId);
    this.emit('signal', message);
  }

  private _setPresence(): void {
    const presenceData: SignalPresenceData = {
      peerId: this._localPeer.peerId,
      metadata: this._localPeer.metadata,
      // Scope tag: on a shared client, other apps' wrappers filter our
      // presence out by this (and we filter theirs).
      __scope: this._options.appName,
    };
    this._roomContext.setPresence(presenceData);
  }
}

File: python/signal/nolag_signal/signal_room.py
from __future__ import annotations

import time
import uuid
from typing import Any, Callable

from .constants import TOPIC_SIGNALING
from .event_emitter import EventEmitter
from .peer_manager import PeerManager
from .types import NoLagSignalOptions, Peer, SignalMessage

try:
    from nolag import EmitOptions, SubscribeOptions
    from nolag.client import Room
except ImportError:
    Room = Any  # type: ignore[assignment,misc]
    EmitOptions = Any  # type: ignore[assignment,misc]
    SubscribeOptions = Any  # type: ignore[assignment,misc]

FilterValue = str | list[str]
"""A single subscription filter value.

A plain string is an OR term: ``["alice", "bob"]`` matches either. A nested
list is an AND group: ``[["alice", "admin"]]`` matches only what was published
tagged with both.
"""


def _merge_filters(existing: list[FilterValue], add: list[str]) -> list[FilterValue]:
    """Merge OR terms into a filter set, preserving AND groups (nested lists)."""
    simple: list[str] = []
    groups: list[list[str]] = []
    for item in existing:
        if isinstance(item, str):
            if item not in simple:
                simple.append(item)
        else:
            groups.append(item)
    for value in add:
        if value not in simple:
            simple.append(value)
    return [*simple, *groups]


def _without_filters(existing: list[FilterValue], remove: list[str]) -> list[FilterValue]:
    """Drop OR terms from a filter set. AND groups are left untouched."""
    drop = set(remove)
    return [item for item in existing if not (isinstance(item, str) and item in drop)]


class SignalRoom(EventEmitter):
    """
    Per-room signaling context.

    Events:
        signal(message: SignalMessage) — Incoming signal for the local peer
        peer_joined(peer: Peer) — Remote peer joined this room
        peer_left(peer: Peer) — Remote peer left this room
    """

    def __init__(
        self,
        name: str,
        room_context: Room,
        local_peer: Peer,
        options: NoLagSignalOptions,
        log: Callable[..., None],
    ) -> None:
        super().__init__()
        self.name = name
        self._room = room_context
        self._local_peer = local_peer
        self._options = options
        self._log = log
        self._peer_manager = PeerManager(local_peer.actor_token_id)
        self._message_handler: Callable[..., None] | None = None
        self._filters: list[FilterValue] = []

    # -- Public: signaling methods --

    async def send_offer(
        self, to_peer_id: str, offer: dict[str, Any], filter: str | None = None
    ) -> None:
        await self.signal(to_peer_id, "offer", offer, filter=filter)

    async def send_answer(
        self, to_peer_id: str, answer: dict[str, Any], filter: str | None = None
    ) -> None:
        await self.signal(to_peer_id, "answer", answer, filter=filter)

    async def send_ice_candidate(
        self, to_peer_id: str, candidate: dict[str, Any], filter: str | None = None
    ) -> None:
        await self.signal(to_peer_id, "ice-candidate", candidate, filter=filter)

    async def send_bye(self, to_peer_id: str, filter: str | None = None) -> None:
        await self.signal(to_peer_id, "bye", {}, filter=filter)

    async def signal(
        self,
        to_peer_id: str,
        signal_type: str,
        payload: dict[str, Any],
        filter: str | None = None,
    ) -> None:
        """Send a signal to a peer.

        By default this broadcasts to the room and peers discard messages not
        addressed to them. Pass ``filter=to_peer_id`` to have the server do the
        addressing instead, so only that peer is sent the signal.
        """
        message = {
            "id": str(uuid.uuid4()),
            "type": signal_type,
            "fromPeerId": self._local_peer.peer_id,
            "toPeerId": to_peer_id,
            "payload": payload,
            "timestamp": time.time() * 1000,
        }
        self._log(f"[{self.name}] Sending {signal_type} to {to_peer_id[:8]}")
        opts = EmitOptions(echo=False, filter=filter) if filter else EmitOptions(echo=False)
        await self._room.emit(TOPIC_SIGNALING, message, opts)

    # -- Public: filters --

    @property
    def local_peer_id(self) -> str:
        """This peer's own id — the value to filter on for directed signals."""
        return self._local_peer.peer_id

    @property
    def filters(self) -> list[FilterValue]:
        """The filter values currently applied to this room's signaling."""
        return list(self._filters)

    async def set_filters(self, values: list[FilterValue]) -> None:
        """Replace this room's signaling filters.

        Only signals published with one of ``values`` are delivered. Set your
        own peer id to receive just the signals addressed to you, and send with
        ``filter=to_peer_id`` so the server does the addressing rather than
        every peer discarding other peers' traffic.

        A filtered peer stops receiving unfiltered room broadcasts, so switch
        the whole room over together. An empty list restores the wildcard
        subscription, which receives everything.
        """
        self._filters = list(values)
        await self._room.set_filters(TOPIC_SIGNALING, list(values))

    async def add_filters(self, values: list[str]) -> None:
        """Add filter values to the existing set. Existing AND groups are kept."""
        await self.set_filters(_merge_filters(self._filters, values))

    async def remove_filters(self, values: list[str]) -> None:
        """Remove filter values. Removing the last one restores the wildcard."""
        await self.set_filters(_without_filters(self._filters, values))

    # -- Public: peer queries --

    def get_peers(self) -> list[Peer]:
        return self._peer_manager.get_all()

    def get_peer(self, peer_id: str) -> Peer | None:
        return self._peer_manager.get_peer(peer_id)

    # -- Internal: called by NoLagSignal --

    async def _subscribe(self, filters: list[FilterValue] | None = None) -> None:
        self._filters = list(filters) if filters else []
        self._message_handler = self._handle_incoming_signal
        self._room.on(TOPIC_SIGNALING, self._message_handler)
        if self._filters:
            await self._room.subscribe(
                TOPIC_SIGNALING, SubscribeOptions(filters=list(self._filters))
            )
        else:
            await self._room.subscribe(TOPIC_SIGNALING)

    async def _activate(self, client: Any) -> None:
        """Set presence and fetch initial room members."""
        await self._set_presence(client)
        presences = client.get_all_presence()
        for ap in presences:
            actor_id = ap.actor_token_id
            presence_data = ap.presence or {}
            peer = self._peer_manager.add_from_presence(actor_id, presence_data, ap.joined_at)
            if peer is not None:
                self._log(f"[{self.name}] Initial peer: {peer.peer_id[:8]}")
                self.emit("peer_joined", peer)

    async def _update_local_presence(self, client: Any) -> None:
        await self._set_presence(client)

    def _handle_presence_join(self, actor_token_id: str, presence_data: dict[str, Any]) -> None:
        peer = self._peer_manager.add_from_presence(actor_token_id, presence_data)
        if peer is not None:
            self._log(f"[{self.name}] Peer joined: {peer.peer_id[:8]}")
            self.emit("peer_joined", peer)

    def _handle_presence_leave(self, actor_token_id: str) -> None:
        peer = self._peer_manager.remove_by_actor_id(actor_token_id)
        if peer is not None:
            self._log(f"[{self.name}] Peer left: {peer.peer_id[:8]}")
            self.emit("peer_left", peer)

    def _handle_presence_update(self, actor_token_id: str, presence_data: dict[str, Any]) -> None:
        self._peer_manager.add_from_presence(actor_token_id, presence_data)

    async def _cleanup(self) -> None:
        if self._message_handler:
            self._room.off(TOPIC_SIGNALING, self._message_handler)
            self._message_handler = None
        await self._room.unsubscribe(TOPIC_SIGNALING)
        self._peer_manager.clear()
        self.remove_all_listeners()

    # -- Private --

    def _handle_incoming_signal(self, data: Any, meta: Any = None) -> None:
        if not isinstance(data, dict):
            return
        to_peer_id = data.get("toPeerId")
        if to_peer_id != self._local_peer.peer_id:
            return

        message = SignalMessage(
            id=data.get("id", ""),
            type=data.get("type", ""),
            from_peer_id=data.get("fromPeerId", ""),
            to_peer_id=to_peer_id,
            payload=data.get("payload", {}),
            timestamp=data.get("timestamp", 0),
        )
        self._log(f"[{self.name}] Received {message.type} from {message.from_peer_id[:8]}")
        self.emit("signal", message)

    async def _set_presence(self, client: Any) -> None:
        presence_data: dict[str, Any] = {"peerId": self._local_peer.peer_id}
        if self._options.metadata:
            presence_data["metadata"] = self._options.metadata
        await client.set_presence(presence_data)
